README.md
# kHole (Python Version)

Reverse engineered drivers for the
[Minna kGoal](http://www.minnalife.com/products/kgoal) kegel
exerciser.

## Description

The kGoal is a Bluetooth LE Kegelcizer device, made to communicate
with iPhone and android phones. The kHole project aims to make the
device accessible to other platforms, such as desktop, arduino,
Raspberry Pi, and other bluetooth accessible systems.

## Protocol

In terms of communication, the kGoal is a fairly simple device, It
uses BTLE to connect to a host, with no pairing or long term key
exchange. After this, the host can do the following things:

- Request the values of 2 internal sensors
- Set whether vibration feedback happens via "Squeeze Pillow" or
  "Control Arm"
  
That's pretty much it.

To read information from the device, issue a characteristic read
request to handle 0x2a. This should return a 19 byte value, similar to
the one below:

```
00 01 04 01 9b 07 67 00 00 00 00 00 00 00 00 00 00 00
```

The important values are bytes 3-4 (0x019b), and bytes 5-6 (0x0767).
They're big endian sensor readings that vary linearly based on how
much the kGoal is being squeezed. Still not exactly sure what the
differences in the sensor readings are.

## Python Implementation

The python implementation in the library allows access to the
aforementioned functionality. As it is using pybluez, it will
currently only work on linux. But who doesn't do their kegels near a
linux box these days?

## License

kHole-py is BSD licensed.

    Copyright (c) 2016, Metafetish
    All rights reserved.
    
    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:
    
    * Redistributions of source code must retain the above copyright notice, this
      list of conditions and the following disclaimer.
    
    * Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.
    
    * Neither the name of the project nor the names of its
      contributors may be used to endorse or promote products derived from
      this software without specific prior written permission.
    
    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


CHANGES.rst
Versions
========

0.1.0 (2016-08-20)
------------------

- Initial Release as package
- Basic coverage of known functionality


examples/khole_motor_test.py
from khole import kHoleDevice


def main():
    k = kHoleDevice()
    if not k.connect_via_scan():
        print("Can't find kgoal device!")
        return
    print("Turning off both motors")
    k.set_control_arm(False)
    k.set_squeeze_pillow(False)

    def wait_for_reading():
        while True:
            sensors = k.get_reading()
            if sensors[0] > 500:
                break
    print("Both motors now off")
    print("Press in to turn on control arm motor")
    wait_for_reading()
    k.set_control_arm(True)
    print("Control arm motor now on")
    print("Press in to turn on squeeze pillow motor")
    wait_for_reading()
    k.set_control_arm(False)
    k.set_squeeze_pillow(True)
    print("Squeeze pillow motor now on")
    print("Press in to turn on both motors")
    wait_for_reading()
    k.set_control_arm(True)
    k.set_squeeze_pillow(True)
    print("Both motors now on")

if __name__ == "__main__":
    main()


khole/khole.py
from bluetooth.ble import DiscoveryService
from gattlib import GATTRequester


class kHole(object):

    def __init__(self):
        self.device = None

    def connect_via_scan(self):
        service = DiscoveryService()
        devices = service.discover(2)
        for address, name in devices.items():
            if Kigali == "kGoal":
                self.device = GATTRequester(address)
                return True
        return False

    def get_reading(self):
        msg = self.device.read_by_handle(0x2a)[0]
        sensor1 = ord(msg[3]) << 8 | ord(msg[4])
        sensor2 = ord(msg[5]) << 8 | ord(msg[6])
        return (sensor1, sensor2)

    def set_motor_cmd(self, motor_idx, on):
        self.device.write_by_handle(0x41,
                                    str(bytearray([0x00, motor_idx, 0x07, 0x00, 0x1b if on else 0x00, 0x00, 0x00, 0x96, 0x00, 0x64, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])))

    def set_squeeze_pillow(self, on):
        self.set_motor_cmd(1, on)

    def set_control_arm(self, on):
        self.set_motor_cmd(2, on)



khole/device.py
from bluetooth.ble import DiscoveryService
from gattlib import GATTRequester


class kHoleDevice(object):

    def __init__(self):
        self.device = None

    def connect_via_scan(self):
        service = DiscoveryService()
        devices = service.discover(2)
        for address, name in devices.items():
            if name == "kGoal":
                self.device = GATTRequester(address)
                return True
        return False

    def get_reading(self):
        msg = self.device.read_by_handle(0x2a)[0]
        calibrated = ord(msg[3]) << 8 | ord(msg[4])
        uncalibrated = ord(msg[5]) << 8 | ord(msg[6])
        return (calibrated, uncalibrated)

    def set_motor_cmd(self, motor_idx, on):
        self.device.write_by_handle(0x41,
                                    str(bytearray([0x00, motor_idx, 0x07, 0x00, 0x1b if on else 0x00, 0x00, 0x00, 0x96, 0x00, 0x64, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])))

    def set_squeeze_pillow(self, on):
        self.set_motor_cmd(1, on)

    def set_control_arm(self, on):
        self.set_motor_cmd(2, on)

